File: sorting-algorithms/quick-sort/JS/quick-sort.js
function partition(arr, low, high) {
  let pivot = arr[low];
  let i = low;
  let j = high;

  while (i <= j) {
    while (arr[i] <= pivot) ++i;
    while (arr[j] > pivot) --j;

    if (i < j) {
      [arr[i], arr[j]] = [arr[j], arr[i]];
    }
  }

  [arr[j], arr[low]] = [arr[low], arr[j]];

  return j;
}

function quickSort(arr, low = 0, high = arr.length - 1) {
  if (low >= high) return;

  let pi = partition(arr, low, high);
  quickSort(arr, low, pi - 1);
  quickSort(arr, pi + 1, high);
}

let arr = [5, 4, 3, 2, 1];
quickSort(arr);
console.log(arr);

File: sorting-algorithms/quick-sort/C/quick-sort.c
#include <stdio.h>

void swap(int *, int *);
int partition(int *, int, int);
void quickSort(int *, int, int);

int main()
{

    int arr[5] = {5, 4, 3, 2, 1};
    quickSort(arr, 0, 4);

    for (int i = 0; i < 5; ++i)
    {
        printf("%d ", arr[i]);
    }

    return 0;
}

void swap(int *a, int *b)
{
    int tmp = *a;
    *a = *b;
    *b = tmp;
}

int partition(int *arr, int low, int high)
{
    int pivot = arr[low];
    int i = low;
    int j = high;

    while (i <= j)
    {
        while (arr[i] <= pivot)
            ++i;
        while (arr[j] > pivot)
            --j;

        if (i < j)
        {
            swap(&arr[i], &arr[j]);
        }
    }

    swap(&arr[low], &arr[j]);

    return j;
}

void quickSort(int *arr, int low, int high)
{
    if (low >= high)
        return;

    int pi = partition(arr, low, high);
    quickSort(arr, low, pi - 1);
    quickSort(arr, pi + 1, high);
}